lib/api/include/irods/ticketAPI.hpp

```diff
@@ -0,0 +1,414 @@
+    {
+        const std::string_view user_name;
+    };
+
+    struct groups_constraint
+    {
+        const std::string_view group_name;
+    };
+
+    struct hosts_constraint
+    {
+        const std::string_view host_name;
+    };
+
+    struct ticket_property_constraint
+    {
+        const ticket_property desired_property;
+        int value_of_property = -1;
+    };
+
+    template <typename T>
+    void addTicketConstraints(RxComm& conn, const std::string_view ticket_name, const T& constraints)
+    {
+        std::string_view command = "mod";
+        std::string_view command_modifier1 = "add";
+        std::string_view command_modifier2 = "";
+        std::string_view command_modifier3 = "";
+        std::string_view command_modifier4 = "";
+        if constexpr (std::is_same_v<users_constraint, T>) {
+            command_modifier2 = "user";
+            command_modifier3 = constraints.user_name;
+        }
+        else if constexpr (std::is_same_v<groups_constraint, T>) {
+            command_modifier2 = "group";
+            command_modifier3 = constraints.group_name;
+        }
+        else if constexpr (std::is_same_v<hosts_constraint, T>) {
+            command_modifier2 = "host";
+            command_modifier3 = constraints.host_name;
+        }
+        else {
+            throw std::invalid_argument("Wrong type object given");
+        }
+
+        execute_ticket_operation(conn,
+                                 command,
+                                 ticket_name,
+                                 command_modifier1,
+                                 command_modifier2,
+                                 command_modifier3,
+                                 command_modifier4,
+                                 false);
+    }
+    template <typename T>
+    void addTicketConstraints(RxComm& conn, const int ticket_id, const T& constraints)
+    {
+        addTicketConstraints(conn, std::to_string(ticket_id), constraints);
+    }
+
+    template <typename T>
+    void addTicketConstraints(admin_tag, RxComm& conn, const std::string_view ticket_name, const T& constraints)
+    {
+        std::string_view command = "mod";
+        std::string_view command_modifier1 = "add";
+        std::string_view command_modifier2 = "";
+        std::string_view command_modifier3 = "";
+        std::string_view command_modifier4 = "";
+        if constexpr (std::is_same_v<users_constraint, T>) {
+            command_modifier2 = "user";
+            command_modifier3 = constraints.user_name;
+        }
+        else if constexpr (std::is_same_v<groups_constraint, T>) {
+            command_modifier2 = "group";
+            command_modifier3 = constraints.group_name;
+        }
+        else if constexpr (std::is_same_v<hosts_constraint, T>) {
+            command_modifier2 = "host";
+            command_modifier3 = constraints.host_name;
+        }
+        else {
+            throw std::invalid_argument("Wrong type object given");
+        }
```

Comment:
This could be a function.

Reply:
I have to do this. Currently, added single line calls after each if, but I will attempt to change it to be a standalone fn.

Reply:
@alanking This can be marked as resolved too.